lib/streamlit/runtime/caching/cache_utils.py

```diff
@@ -378,24 +165,19 @@ def get_or_create_cached_value():
                 result = cache.read_result(value_key)
                 _LOGGER.debug("Cache hit: %s", func)
 
-                replay_result_messages(result, cached_func.cache_type, func)
+                replay_cached_messages(result, cached_func.cache_type, func)
 
                 return_value = result.value
 
             except CacheKeyNotFoundError:
                 _LOGGER.debug("Cache miss: %s", func)
 
-                with cached_func.warning_call_stack.calling_cached_function(func):
-                    with cached_func.message_call_stack.calling_cached_function():
-                        with cached_func.warning_call_stack.maybe_allow_widgets(
-                            cached_func.allow_widgets
-                        ):
-                            with cached_func.message_call_stack.maybe_allow_widgets(
-                                cached_func.allow_widgets
-                            ):
-                                return_value = func(*args, **kwargs)
-
-                messages = cached_func.message_call_stack._most_recent_messages
+                with cached_func.cached_message_replay_ctx.calling_cached_function(
+                    func, cached_func.allow_widgets
+                ):
+                    return_value = func(*args, **kwargs)
```

Comment:
"warning_call_stack" and "message_call_stack" are now the same object, and `calling_cached_function` takes the "allow_widgets" flag, so we can collapse these 4 context managers down into 1